ironfish/src/network/messages/identify.ts

```diff
@@ -0,0 +1,98 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import bufio from 'bufio'
+import { BigIntUtils } from '../../utils/bigint'
+import { Identity, identityLength } from '../identity'
+import { NetworkMessage, NetworkMessageType } from './networkMessage'
+
+interface CreateIdentifyMessageOptions {
+  agent: string
+  head: Buffer
+  identity: Identity
+  name?: string
+  port: number | null
+  sequence: number
+  version: number
+  work: bigint
+}
+
+export class IdentifyMessage extends NetworkMessage {
+  readonly agent: string
+  readonly head: Buffer
+  readonly identity: Identity
+  readonly name: string
+  readonly port: number
+  readonly sequence: number
+  readonly version: number
+  readonly work: bigint
+
+  constructor({
+    agent,
+    head,
+    identity,
+    name,
+    port,
+    sequence,
+    version,
+    work,
+  }: CreateIdentifyMessageOptions) {
+    super(NetworkMessageType.Identify)
+    this.agent = agent
+    this.head = head
+    this.identity = identity
+    this.name = name || ''
+    this.port = port || 0
+    this.sequence = sequence
+    this.version = version
+    this.work = work
+  }
+
+  serialize(): Buffer {
+    const bw = bufio.write(this.getSize())
+    bw.writeBytes(Buffer.from(this.identity, 'base64'))
+    bw.writeVarString(this.name)
+    bw.writeU16(this.port)
+    bw.writeU16(this.version)
+    bw.writeVarString(this.agent)
+    bw.writeU32(this.sequence)
+    bw.writeHash(this.head)
+    bw.writeVarBytes(BigIntUtils.toBytesLE(this.work))
```

Comment:
This could probably be constrained -- it's a function of the target, I believe, but it'll likely be smaller than the full 256 bits of the target, so I think leaving it as a variable-length field isn't a huge deal.